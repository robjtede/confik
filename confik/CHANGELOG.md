# Changelog

## Unreleased

## 0.10.0

- Include the index of an unexpected secret when one is found in an unkeyed container (such as a `Vec`). Note that this will provide little to no information for unsorted/arbitrarily containers like a `HashSet`.
- `Option`s will now obey `default`s when they have received no explicit configuration.
  - Explicit `null`s, such as in JSON will set the `Option` to `None`, not to the default.
- Containers, such as `Vec` and `HashMap` will now obey `default`s when they have received no explicit configuration.
  - Explicit empty sets will create an empty container.
- Partial configuration will return an error, even if there is a default set at a higher level. E.g.
    ```rust
    struct Data {
      a: usize,
      b: usize
    }

    struct Config {
      #[confik(default = Data { a: 1, b: 2 })]
      data: Data,
    }
    ```

    with configuration:

    ```toml
    [data]
    a = "5"
    ```

    will return an error indicating `b` is missing, instead of ignoring the provided configuration.

## 0.9.0

- Optional crate features no longer have the `with-` prefix, e.g.: `with-uuid` -> `uuid`.

## 0.8.0

- Attributes that receive expressions (`default` and `from`) now need to be unquoted, e.g.:

  ```diff
  - struct Config { #[config(default = "\"Hello World\"") param: String }
  + struct Config { #[config(default = "Hello World") param: String }

  - struct Config { #[config(default = "5_usize") param: usize }
  + struct Config { #[config(default = 5_usize) param: usize }
  ```

## 0.7.0

- Initial release.
