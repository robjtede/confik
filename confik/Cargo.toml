[package]
name = "confik"
version = "0.10.0"
description = "A library for reading application configuration split across multiple sources"
authors = ["Rob Ede <robjtede@icloud.com>"]
keywords = ["parser", "serde", "utility", "config"]
categories = ["config"]
repository = "https://github.com/robjtede/confik"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.66"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

[package.metadata.cargo-machete]
ignored = [
  "humantime_serde", # used in serde macro attributes
  "serde_with",      # used in doctests
]

[features]
default = ["env", "toml"]

# Source types
env = ["dep:envious"]
json = ["dep:serde_json"]
toml = ["dep:toml"]

# Destination types
common = []
chrono = ["dep:chrono"]
rust_decimal = ["dep:rust_decimal"]
url = ["dep:url"]
uuid = ["dep:uuid"]

[dependencies]
confik-macros = "=0.10.0"

cfg-if = "1"
serde = { version = "1", default-features = false, features = ["std", "derive"] }
thiserror = "1.0.41"

envious = { version = "0.2", optional = true }
serde_json = { version = "1", optional = true, default-features = false, features = ["std"] }
toml = { version = "0.7", optional = true, default-features = false, features = ["parse"] }

chrono = { version = "0.4.20", optional = true, default-features = false }
rust_decimal = { version = "1", optional = true, default-features = false, features = ["serde"] }
url = { version = "2", optional = true, default-features = false, features = ["serde"] }
uuid = { version = "1", optional = true, default-features = false, features = ["serde"] }

[dev-dependencies]
assert_matches = "1.5.0"
humantime-serde = "1"
indoc = "2"
serde_with = "3"
temp-env = "0.3"

[[example]]
name = "simple"
required-features = ["toml"]

[[example]]
name = "derives"
required-features = ["toml"]
