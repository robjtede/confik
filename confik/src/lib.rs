//! # `confik`
//!
//! `confik` is a library for reading application configuration split across multiple sources.
//!
//! ## Example
//!
//! Assume that `config.toml` contains:
//!
//! ```toml
//! host=google.com
//! username=root
//! ```
//!
//! and the environment contains:
//!
//! ```bash
//! PASSWORD=hunter2
//! ```
//!
//! then:
//!
//! ```no_run
//! # #[cfg(all(feature = "toml", feature = "env"))]
//! # {
//! use confik::{Configuration, EnvSource, FileSource, TomlSource};
//!
//! #[derive(Debug, PartialEq, Configuration)]
//! struct Config {
//!     host: String,
//!     username: String,
//!
//!     #[confik(secret)]
//!     password: String,
//! }
//!
//! let config = Config::builder()
//!     .override_with(FileSource::new("config.toml"))
//!     .override_with(EnvSource::new().allow_secrets())
//!     .try_build()
//!     .unwrap();
//!
//! assert_eq!(
//!     config,
//!     Config {
//!         host: "google.com".to_string(),
//!         username: "root".to_string(),
//!         password: "hunter2".to_string(),
//!     }
//! );
//! # }
//! ```
//!
//! ## Sources
//!
//! A [`Source`] is any type that can create [`ConfigurationBuilder`]s.
//! This crate implements the following sources:
//!
//! - [`EnvSource`]:
//!   Loads configuration from environment variables using the [`envious`] crate.
//!   Requires the `env` feature. (Enabled by default.)
//! - [`FileSource`]:
//!   Loads configuration from a file, detecting `json` or `toml` files based on the file extension.
//!   Requires the `json` and `toml` feature respectively. (`toml` is enabled by default.)
//! - [`TomlSource`]:
//!   Loads configuration from a TOML string literal.
//!   Requires the `toml` feature. (Enabled by default.)
//! - [`JsonSource`]:
//!   Loads configuration from a JSON string literal.
//!   Requires the `json` feature.
//!
//! ## Secrets
//!
//! Fields annotated with `#[confik(secret)]` will only be read from secure sources.
//! This serves as a runtime check that no secrets have been stored in insecure places such as
//! world-readable files.
//!
//! If a secret is found in an insecure source, an error will be returned.
//! You can opt into loading secrets on a source-by-source basis.
//!
//! ## Foreign Types
//!
//! This crate provides implementations of [`Configuration`] for a number of `std` types and the
//! following third-party crates. Implementations for third-party crates are feature gated.
//!
//! - `with-chrono` - `chrono` 0.4
//! - `with-rust_decimal` - `rust_decimal` 1
//! - `with-url` - `url` 1
//! - `with-uuid` - `uuid` 1
//!
//! ## Macro usage
//!
//! The derive macro is called `Configuration` and is used as normal:
//!
//! ```
//! #[derive(confik::Configuration)]
//! struct Config {
//!     data: usize,
//! }
//! ```
//!
//! ### Forwarding Attributes To `Deserialize`
//!
//! The serde attributes used for customizing a `Deserialize` derive typically are achieved by
//! adding `#[confik(forward_serde(...))` attributes.
//!
//! For example:
//!
//! ```
//! #[derive(confik::Configuration)]
//! struct Config {
//!     #[confik(forward_serde(rename = "other_data"))]
//!     data: usize,
//! }
//! ```
//!
//! ### Defaults
//!
//! Defaults are specified on a per-field basis.
//! * Defaults are used if the data cannot be fully read, even if it is partially read.
//!   E.g., even if `data` in the below example has one value read in, both will be overwritten by
//!   the default.
//!   ```
//!   # #[cfg(feature = "toml")]
//!   # {
//!   use confik::{Configuration, TomlSource};
//!
//!   #[derive(Configuration)]
//!   struct Data {
//!       a: usize,
//!       b: usize,
//!   }
//!
//!   #[derive(Configuration)]
//!   struct Config {
//!       #[confik(default = Data  { a: 1, b: 2 })]
//!       data: Data
//!   }
//!
//!   let toml = r#"
//!       [data]
//!       a = 1234
//!   "#;
//!
//!   let config = Config::builder()
//!       .override_with(TomlSource::new(toml))
//!       .try_build()
//!       .unwrap();
//!   assert_eq!(config.data.a, 1);
//!
//!   let toml = r#"
//!       [data]
//!       a = 1234
//!       b = 4321
//!   "#;
//!
//!   let config = Config::builder()
//!       .override_with(TomlSource::new(toml))
//!       .try_build()
//!       .unwrap();
//!   assert_eq!(config.data.a, 1234);
//!   # }
//!   ```
//! * Defaults can be given by any rust expression, and have [`Into::into`] run over them. E.g.,
//!   ```
//!   const DEFAULT_VALUE: u8 = 4;
//!
//!   #[derive(confik::Configuration)]
//!   struct Config {
//!       #[confik(default = DEFAULT_VALUE)]
//!       a: u32,
//!       #[confik(default = "hello world")]
//!       b: String,
//!       #[confik(default = 5f32)]
//!       c: f32,
//!   }
//!   ```
//! * Alternatively, a default without a given value called [`Default::default`]. E.g.,
//!   ```
//!   use confik::{Configuration};
//!
//!   #[derive(Configuration)]
//!   struct Config {
//!       #[confik(default)]
//!       a: usize
//!   }
//!
//!   let config = Config::builder().try_build().unwrap();
//!   assert_eq!(config.a, 0);
//!   ```
//!
//! ### Handling Foreign Types
//!
//! If there's a foreign type used in your config, then you will not be able to implement
//! [`Configuration`] for it. Instead any type that implements [`Into`] can be used.
//!
//! ```
//! struct ForeignType {
//!     data: usize,
//! }
//!
//! #[derive(confik::Configuration)]
//! struct MyForeignTypeCopy {
//!     data: usize
//! }
//!
//! impl From<MyForeignTypeCopy> for ForeignType {
//!     fn from(copy: MyForeignTypeCopy) -> Self {
//!         Self {
//!             data: copy.data,
//!         }
//!     }
//! }
//!
//! #[derive(confik::Configuration)]
//! struct Config {
//!     #[confik(from = "MyForeignTypeCopy")]
//!     foreign_data: ForeignType
//! }
//! ```
//!
//! ## Macro Limitations
//!
//! ### `Option` Defaulting
//!
//! `Option`s cannot default to anything other than `None`. I.e., the below example ignores the provided default.
//!
//! ```
//! # use confik::Configuration;
//!
//! const DEFAULT_DATA: Option<usize> = Some(5);
//!
//! #[derive(Configuration)]
//! struct Config {
//!     #[confik(default = DEFAULT_DATA)]
//!     data: Option<usize>
//! }
//!
//! let config = Config::builder().try_build().unwrap();
//! assert_eq!(config.data, None);
//! ```
//!
//! This behaviour occurs due to `Option`s needing to have a default value of `None`, as `Option`al
//! configuration shouldn't be required. This defaulting occurs inside the [`ConfigurationBuilder`]
//! implementation of [`Option`] and so happens before the macro can try to default the value.
//!
//! This is in principle fixable by special casing any value with an `Option<...>` type, but this
//! has not been implemented due to the fragility that trying to exact match on the string value of
//! a type in a macro would bring. E.g., a custom type such as `type MyOption = Option<usize>` would
//! then behave differently to using `Option<usize>` directly.
//!
//! ### Custom `Deserialize` Implementations
//!
//! If you're using a custom `Deserialize` implementation, then you cannot use the `Configuration`
//! derive macro. Instead, define the necessary config implementation manually like so:
//!
//! ```rust
//! #[derive(Debug, serde_with::DeserializeFromStr)]
//! enum MyEnum {
//!     Foo,
//!     Bar,
//! };
//!
//! impl std::str::FromStr for MyEnum {
//!     // ...
//! # type Err = String;
//! # fn from_str(_: &str) -> Result<Self, Self::Err> { unimplemented!() }
//! }
//!
//! impl confik::Configuration for MyEnum {
//!     type Builder = Option<Self>;
//! }
//! ```
//!
//! Note that the `Option<Self>` builder type only works for simple types. For more info, see the
//! docs on [`Configuration`] and [`ConfigurationBuilder`].

#![deny(rust_2018_idioms, nonstandard_style, future_incompatible)]
#![cfg_attr(docsrs, feature(doc_auto_cfg))]

use std::{borrow::Cow, ops::Not};

#[doc(hidden)]
pub use confik_macros::*;

use crate::{path::Path, sources::DynSource};

#[doc(hidden)]
pub mod __exports {
    /// Re-export [`Deserialize`] for use in case `serde` is not otherwise used
    /// whilst we are.
    ///
    /// As serde then calls into other serde functions, we need to re-export the whole of serde,
    /// instead of just [`Deserialize`].
    ///
    /// [`Deserialize`]: serde::Deserialize
    pub use serde as __serde;
}

// Enable use of macros inside the crate
#[allow(unused_extern_crates)] // false positive
extern crate self as confik;

mod builder;
#[cfg(feature = "common")]
pub mod common;
mod errors;
mod path;
mod secrets;
mod sources;
mod std_impls;
mod third_party;

#[cfg(feature = "env")]
pub use self::sources::env_source::EnvSource;
#[cfg(feature = "json")]
pub use self::sources::json_source::JsonSource;
#[cfg(feature = "toml")]
pub use self::sources::toml_source::TomlSource;
pub use self::{
    builder::ConfigBuilder,
    errors::Error,
    secrets::{SecretBuilder, UnexpectedSecret},
    sources::{file_source::FileSource, Source},
};

/// Captures the path of a missing value.
#[derive(Debug, Default, thiserror::Error)]
#[error("Missing value for path `{0}`")]
pub struct MissingValue(Path);

impl MissingValue {
    /// Prepends a path segment as we return back up the call-stack.
    #[must_use]
    pub fn prepend(mut self, path_segment: impl Into<Cow<'static, str>>) -> Self {
        self.0 .0.push(path_segment.into());
        self
    }
}

/// Converts the sources, in order, into [`Configuration::Builder`] and
/// [`ConfigurationBuilder::merge`]s them, passing any errors back.
fn build_from_sources<'a, Target, Iter>(sources: Iter) -> Result<Target, Error>
where
    Target: Configuration,
    Iter: IntoIterator<Item = Box<dyn DynSource<Target::Builder> + 'a>>,
{
    sources
        .into_iter()
        // Convert each source to a `Target::Builder`
        .map::<Result<Target::Builder, Error>, _>(|s: Box<dyn DynSource<Target::Builder> + 'a>| {
            let debug = || format!("{:?}", s);
            let res = s.provide().map_err(|e| Error::Source(e, debug()))?;
            if s.allows_secrets().not() {
                res.contains_non_secret_data()
                    .map_err(|e| Error::UnexpectedSecret(e, debug()))?;
            }
            Ok(res)
        })
        // Merge the builders
        .reduce(|first, second| Ok(Target::Builder::merge(first?, second?)))
        // If there was no data then we're missing values
        .ok_or_else(|| Error::MissingValue(MissingValue::default()))??
        .try_build()
        .map_err(Into::into)
}

/// The target to be deserialized from multiple sources.
///
/// This will normally be created by the derive macro which also creates a [`ConfigurationBuilder`]
/// implementation.
///
/// For types with no contents, e.g. empty structs, or simple enums, this can be implemented very
/// easily by specifying only the builder type as `Option<Self>`. For anything more complicated,
/// complete target and builder implementations will be needed.
///
/// # Examples
///
/// ```
/// use confik::Configuration;
///
/// #[derive(serde::Deserialize)]
/// enum MyEnum { A, B, C }
///
/// impl Configuration for MyEnum {
///     type Builder = Option<Self>;
/// }
/// ```
pub trait Configuration: Sized {
    /// The builder that accumulates the deserializations.
    type Builder: ConfigurationBuilder<Target = Self>;

    /// Creates an instance of [`ConfigBuilder`] tied to this type.
    #[must_use]
    fn builder<'a>() -> ConfigBuilder<'a, Self> {
        ConfigBuilder::<Self>::default()
    }
}

/// A builder for a multi-source config deserialization.
///
/// This will almost never be implemented manually, instead being derived.
///
/// Builders must implement [`Default`] so that if the structure is nested in another then it being
/// missing is not an error.
/// For trivial cases, this is solved by using an `Option<Configuration>`.
/// See the worked example on [`Configuration`].
pub trait ConfigurationBuilder: Default + serde::de::DeserializeOwned {
    /// The target that will be converted into. See [`Configuration`].
    type Target;

    /// Combines two builders recursively, preferring `self`'s data, if present.
    #[must_use]
    fn merge(self, other: Self) -> Self;

    /// This will probably delegate to `TryInto` but allows it to be implemented for types foreign
    /// to the library.
    fn try_build(self) -> Result<Self::Target, MissingValue>;

    /// Called recursively on each field, aiming to hit all [`SecretBuilder`]s. This is only called
    /// when [`Source::allows_secrets`] is `false`.
    ///
    /// If any data is present then `Ok(true)` is returned, unless the data is wrapped in a
    /// [`SecretBuilder`] in which case [`UnexpectedSecret`] is passed, which will then be built
    /// into the path to the secret data.
    fn contains_non_secret_data(&self) -> Result<bool, UnexpectedSecret>;
}

/// Implementations for trivial types via `Option`.
///
/// This can also be used for user types, such as an `enum` with no variants containing fields. See
/// the worked example on [`Configuration`].
impl<T> ConfigurationBuilder for Option<T>
where
    T: serde::de::DeserializeOwned + Configuration,
{
    type Target = T;

    fn merge(self, other: Self) -> Self {
        self.or(other)
    }

    fn try_build(self) -> Result<Self::Target, MissingValue> {
        self.ok_or_else(|| MissingValue(Path::new()))
    }

    /// Should not have an `Option` wrapping a secret as `<Option<T> as ConfigurationBuilder` is
    /// used for terminal types, therefore the `SecretBuilder` wrapping would be external to it.
    fn contains_non_secret_data(&self) -> Result<bool, UnexpectedSecret> {
        Ok(self.is_some())
    }
}
