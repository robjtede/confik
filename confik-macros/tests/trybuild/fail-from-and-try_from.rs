//! Check that the `from` attribute works
use confik::{Configuration, TomlSource};

#[derive(Debug, Configuration, PartialEq, Eq)]
struct Config {
    #[confik(try_from = A, from = B)]
    param: String,
}

#[derive(Debug, Default, serde::Deserialize, confik::Configuration)]
struct A(usize);

#[derive(Debug, Default, serde::Deserialize, confik::Configuration)]
struct B(usize);

#[derive(Debug)]
struct E;

impl std::fmt::Display for E {
    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Ok(())
    }
}

impl std::error::Error for E {}

impl TryFrom<A> for String {
    type Error = E;

    fn try_from(_: A) -> Result<Self, Self::Error> {
        Ok(String::from("Hello world"))
    }
}

impl From<B> for String {
    fn from(_: B) -> Self {
        String::from("Hello world")
    }
}

fn main() {
    let config = Config::builder()
        .override_with(TomlSource::new("param = 5"))
        .try_build()
        .expect("Failed to build with no required data");
    assert_eq!(
        Config {
            param: String::from("Hello world")
        },
        config
    );
}
