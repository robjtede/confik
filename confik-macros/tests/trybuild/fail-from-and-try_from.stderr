error: Cannot support both `try_from` and `from` confik attributes
 --> tests/trybuild/fail-from-and-try_from.rs:6:25
  |
6 |     #[confik(try_from = A, from = B)]
  |                         ^

error: Cannot support both `try_from` and `from` confik attributes
 --> tests/trybuild/fail-from-and-try_from.rs:6:35
  |
6 |     #[confik(try_from = A, from = B)]
  |                                   ^

error[E0599]: no function or associated item named `builder` found for struct `Config` in the current scope
  --> tests/trybuild/fail-from-and-try_from.rs:42:26
   |
5  | struct Config {
   | ------------- function or associated item `builder` not found for this struct
...
42 |     let config = Config::builder()
   |                          ^^^^^^^ function or associated item not found in `Config`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `builder`, perhaps you need to implement it:
           candidate #1: `Configuration`
